File: packages/wazeko-types/src/events.ts
import { Jid } from "./jid.js";
import { Message, MessageId, MessageStatus } from "./message.js";
import { GroupMetadata } from "./group.js";

export type ConnectionState =
  | "disconnected"
  | "connecting"
  | "authenticating"
  | "connected"
  | "reconnecting"
  | "logged_out";

export interface QrCodeEvent {
  raw: string;
  attempts: number;
  timeoutSeconds: number;
}

export interface PairingCodeEvent {
  code: string;
  expiresInSeconds: number;
}

export interface MessageReceiptEvent {
  messageId: MessageId;
  chat: Jid;
  sender?: Jid;
  status: MessageStatus;
  timestamp: number;
}

export type WazekoEventMap = {
  "connection.update": ConnectionState;
  "qr": QrCodeEvent;
  "pairing.code": PairingCodeEvent;
  "authenticated": { userJid: Jid };
  "message": Message;
  "message.receipt": MessageReceiptEvent;
  "message.delete": { messageId: MessageId; chat: Jid };
  "group.update": GroupMetadata;
  "presence.update": { jid: Jid; presence: string };
  "error": Error;
  "disconnect": { reason: string };
};

export type WazekoEventName = keyof WazekoEventMap;

File: crates/wazeko-types/src/events.rs
use crate::group::GroupMetadata;
use crate::jid::Jid;
use crate::message::{Message, MessageId, MessageStatus};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ConnectionState {
    Disconnected,
    Connecting,
    Authenticating,
    Connected,
    Reconnecting,
    LoggedOut,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QrCodeEvent {
    pub raw: String,
    pub attempts: u32,
    pub timeout_seconds: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PairingCodeEvent {
    pub code: String,
    pub expires_in_seconds: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MessageReceiptEvent {
    pub message_id: MessageId,
    pub chat: Jid,
    pub sender: Option<Jid>,
    pub status: MessageStatus,
    pub timestamp: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Event {
    ConnectionUpdate(ConnectionState),
    Qr(QrCodeEvent),
    PairingCode(PairingCodeEvent),
    Authenticated { user_jid: Jid },
    Message(Message),
    MessageReceipt(MessageReceiptEvent),
    MessageDeleted { message_id: MessageId, chat: Jid },
    GroupUpdate(GroupMetadata),
    PresenceUpdate { jid: Jid, presence: String },
    Error(String),
    Disconnected { reason: String },
}
